designs/conditional-controllers.md

```diff
@@ -0,0 +1,204 @@
+steps.
+
+The first goal of enabling the possibility of starting/stopping/restarting
+controllers is addressed in two parts, first where a mechanism is built to
+enable the removal of informers, and second where we expose this mechanism and
+enable the restarting of controllers
+
+The second goal of a solution for running controllers conditionally is addressed
+by creating a wrapper around a controller called a ConditionalController that
+within its Start() function, it polls the discovery doc for the existence of the
+watched object and starts/stops/restarts the underlying controller as necessary 
+
+The third goal of an ergonomic mechanism to use ConditionalControllers is a
+small modification to the controller builder to enable running a controller as a
+ConditionalController.
+
+###### Proof of concept
+A proof of concept PR exists at
+[#1180](https://github.com/kubernetes-sigs/controller-runtime/pull/1180)
+
+Each commit maps to one step in the proposal and can loosely be considered going
+from most necessary to least.
+
+### Informer Removal [Pre-req]
+
+This proposal assumes a mechanism exists for removing individual informers. We
+are agnostic to how this is done, but the current proposal is built on top of
+Shomron’s proposed implementation of individual informer removal
+[#936](https://github.com/kubernetes-sigs/controller-runtime/pull/936).
+
+
+Risks/mitigations and alternatives are discussed in the linked PR as well as the
+corresponding issue
+[#935](https://github.com/kubernetes-sigs/controller-runtime/issues/935). 
+
+We are currently looking into whether that PR can be distilled further into a more
+minimal solution.
+
+### Minimal hooks
+
+Given that a mechanism exists to remove individual informers, the next step is
+to expose this removal functionality and enable safely
+starting/stopping/restarting controllers and their caches.
+
+The proposal to do this is:
+1. Expose the `informerCache.Remove()` functionality on the cache interface.
+2. Expose the ability to reset the internal controller’s `Started` field.
+3. Expose a field on the internal controller to determine whether to `saveWatches`
+or not and use this field to not empty the controller’s `startWatches` when the
+controller is stopped.
+
+#### Risks and Mitigations
+
+* [#1139](https://github.com/kubernetes-sigs/controller-runtime/pull/1139) discusses why
+the ability to start a controller more than once was taken away. It's a little
+unclear what effect explicitly enabling multiple starts in the case of
+conditional controllers will hae on the number of workers and workqueues
+relative to expectations and metrics.
+* [#1163](https://github.com/kubernetes-sigs/controller-runtime/pull/1163) discusses the
+memory leak caused by no clearing out the watches internal slice. A possible
+mitigation is to clear out the slices upon ConditionalController shutdown.
+
+#### Alternatives
+
+* Instead of exposing ResetStart and SaveWatches on the internal controller struct
```

Comment:
Can we mention the option of a CRD-controller here that starts/stops other controllers based on the presence/absence of crds?

Reply:
done, lemme if I've captured the pros/cons accurately